kernel: Add block header support and validation

yuvicc opened the pull request:
Adds a new `btck_BlockHeader` type and associated functions to create, access, and validate block headers. Block headers will have their own type (`btck_BlockHeader`) that can be created from raw data, copied, and queried for all the standard header fields (hash, prev hash, timestamp, bits, version, nonce). We can also extract headers from full blocks or block tree entries.

The first commit here refactors `BlockValidationState` to use Handle/View pattern so external code can own them, which is required for the header processing in the API.

 #### New Block Header API

  - **`btck_BlockHeader` type**: Opaque handle for block headers
  - **Header methods**:
    - `btck_block_header_create()`: Create header from 80-byte serialized data
    - `btck_block_header_copy()`: Copy block headers
    - `btck_block_header_destroy()`: Destroy header object
    - `btck_block_header_get_hash()`: Calculate block hash
    - `btck_block_header_get_prev_hash()`: Get previous block hash
    - `btck_block_header_get_timestamp()`: Get block timestamp
    - `btck_block_header_get_bits()`: Get difficulty target (compact format)
    - `btck_block_header_get_version()`: Get block version
    - `btck_block_header_get_nonce()`: Get proof-of-work nonce
    
    - `btck_block_get_header()`: Extract header from a full block
    - `btck_block_tree_entry_get_block_header()`: Get header associated with a block tree entry
  
  - **Header Processing Methods:**
    - **`btck_chainstate_manager_process_block_header()`**: Validates and processes a block header without requiring the full block. This performs proof-of-work verification, timestamp validation, and updates the internal chain state.
    - **`btck_chainstate_manager_get_best_entry()`**: Returns the block tree entry with the most cumulative proof-of-work.

Why `btck_chainstate_manager_get_best_entry()` is included alongside header validation? Just as we have logic to get the tip for block validation (so you can request more blocks extending your best from your peers), we need the equivalent for header validation. To make header validation worthwhile, knowing what the best current header is seems useful—it tells you what headers to request next from peers.



  ### Testing

  Added tests in `test_kernel.cpp` that cover creating headers from raw data, extracting all header fields, and processing headers through the chainstate manager.
  
  CC sedited

TheCharlatan commented on src/bitcoin-chainstate.cpp:
```diff
@@ -44,17 +44,17 @@ class TestValidationInterface : public ValidationInterface
     std::optional<std::string> m_expected_valid_block = std::nullopt;
 
-    void BlockChecked(const Block block, const BlockValidationState state) override
+    void BlockChecked(Block block, const BlockValidationStateView view) override
```
I think the variable name `state` should just be kept.

TheCharlatan commented on src/test/kernel/test_kernel.cpp:
```diff
@@ -574,6 +574,14 @@ BOOST_AUTO_TEST_CASE(btck_context_tests)
+{
+    BlockHeader header_0{hex_string_to_byte_vec("00e07a26beaaeee2e71d7eb19279545edbaf15de0999983626ec00000000000000000000579cf78b65229bfb93f4a11463af2eaa5ad91780f27f5d147a423bea5f7e4cdf2a47e268b4dd01173a9662ee")};
+    BOOST_CHECK_EQUAL(byte_span_to_hex_string_reversed(header_0.Hash().ToBytes()), "00000000000000000000325c7e14a4ee3b4fcb2343089a839287308a0ddbee4f");
+    BlockHeader header_1{hex_string_to_byte_vec("00c00020e7cb7b4de21d26d55bd384017b8bb9333ac3b2b55bed00000000000000000000d91b4484f801b99f03d36b9d26cfa83420b67f81da12d7e6c1e7f364e743c5ba9946e268b4dd011799c8533d")};
```
Can you add `BOOST_CHECK_THROW(BlockHeader{hex_string_to_byte_vec("00")}, std::runtime_error);`?

TheCharlatan commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1259,3 +1302,64 @@ int btck_chain_contains(const btck_Chain* chain, const btck_BlockTreeEntry* entr
+    if (raw_block_header_len != 80) {
+        LogError("invalid header length %zu (expected 80)", raw_block_header_len);
+        return nullptr;
+    }
```
I wouldn't add this length check here and would instead just rely on the exception. We do the same thing in the rpc too: `submitheader` just calls `DecodeHexBlockHeader`, which doesn't have a length check either.

stringintech replied:
After taking another look, it makes sense to replicate the behavior of `submitheader` here (which accepts trailing bytes). Looking across the functional tests, there are multiple usages of the `submitheader` RPC passing whole block data instead of just the header.

I was initially confused by `DecodeHexTx` which IS strict about unconsumed bytes, but notably `DecodeHexBlk` (for full blocks) also doesn't enforce this check. The strictness in `DecodeTx` appears to be related to differentiating between witness vs non-witness transaction formats.

TheCharlatan commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1184,6 +1238,11 @@ BITCOINKERNEL_API void btck_block_destroy(btck_Block* block);
+/**
+ * Create a new block validation state.
+ */
+BITCOINKERNEL_API btck_BlockValidationState* btck_block_validation_state_create();
```
This (and `btck_block_validation_state_copy` needs `BITCOINKERNEL_WARN_UNUSED_RESULT`

sedited reviewed (commented):
Approach ACK

yancyribbens commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1225,6 +1258,16 @@ int btck_chainstate_manager_process_block(
+    btck_BlockValidationState* state)
+{
+    auto& chainman = btck_ChainstateManager::get(chainstate_manager).m_chainman;
+    auto result = chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, true, btck_BlockValidationState::get(state), nullptr);
```
Isn't this something the fuzz harness can help answer?

sedited replied:
I think that is fair. We are doing a bunch of disk reads that may throw when processing new blocks, so adding a try-catch there too seems like a good idea.

stickies-v reviewed (commented):
Approach ACK 86958ee668a950a03d08ef188f2de27137e89b33

I think the commit can be broken up a bit further, e.g. validationstate logic and chainman logic can be separate I think?

yuvicc commented on src/kernel/bitcoinkernel.h:
```diff
@@ -904,6 +914,16 @@ BITCOINKERNEL_API void btck_context_destroy(btck_Context* context);
 
+/**
+ * @brief Return the block header associated with this entry. The returned
+ * header is not owned and depends on the lifetime of the block.
```
Correct, Thanks.

yuvicc committed:
kernel: Add Handle/View pattern for BlockValidationState

Add C API functions for managing BlockValidationState lifecycle:
  - btck_block_validation_state_create()
  - btck_block_validation_state_copy()
  - btck_block_validation_state_destroy()

Introduce BlockValidationStateApi<> template to share common getter methods between BlockValidationState (Handle) and BlockValidationStateView (View) classes in the C++ wrapper. This enables external code to create and own BlockValidationState objects needed for the new process_block_header() API.

Co-authored-by: TheCharlatan <seb.kung@gmail.com>

TheCharlatan commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1588,6 +1641,92 @@ BITCOINKERNEL_API void btck_block_hash_destroy(btck_BlockHash* block_hash);
+ *
+ * @param[in] raw_block_header        Non-null, serialized header data (80 bytes)
+ * @param[in] raw_block_header_len    Length of serialized header (must be 80)
+ * @return                      Block Header, or null on error. Must be destroyed with btck_block_header_destroy().
```
The indentation should be fixed here.

yuvicc commented:
Thanks for the review @TheCharlatan 
- Removed some unnecessary comments and indentation
- Added inline doc [comment](https://github.com/bitcoin/bitcoin/pull/33822#discussion_r2528773169)

stringintech commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1017,6 +1036,29 @@ BITCOINKERNEL_API void btck_chainstate_manager_options_destroy(btck_ChainstateMa
+ * @param[out] block_validation_state  The result of the header validation.
+ * @return              0 if header processing completed successfully, non-zero on error.
+ */
+BITCOINKERNEL_API int BITCOINKERNEL_WARN_UNUSED_RESULT btck_chainstate_manager_process_block_header(
```
Indentation can be fixed here.

stringintech commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1568,6 +1641,92 @@ BITCOINKERNEL_API void btck_block_hash_destroy(btck_BlockHash* block_hash);
+ * @return              Previous block hash (view, do not destroy)
+ */
+BITCOINKERNEL_API const btck_BlockHash* BITCOINKERNEL_WARN_UNUSED_RESULT btck_block_header_get_prev_hash(
+    const btck_BlockHeader* header) BITCOINKERNEL_ARG_NONNULL(1);
```
We have usually documented the non-owned pointers this way: `The returned X is unowned and only valid for the lifetime of Y.`
```suggestion
/**
 * @brief Get the previous block hash from block header. The returned hash is unowned and only valid for the lifetime of the block header.
 *
 * @param[in] header    Non-null header
 * @return              Previous block hash
 */
BITCOINKERNEL_API const btck_BlockHash* BITCOINKERNEL_WARN_UNUSED_RESULT btck_block_header_get_prev_hash(
    const btck_BlockHeader* header) BITCOINKERNEL_ARG_NONNULL(1);
```

stringintech reviewed (commented):
Approach ACK

yuvicc commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -854,6 +855,21 @@ const btck_BlockTreeEntry* btck_block_tree_entry_get_previous(const btck_BlockTr
 
+btck_BlockValidationState* btck_block_validation_state_create()
+{
+    return btck_BlockValidationState::create(BlockValidationState{});
```
Correct.

yuvicc commented:
Thanks for the review @stringintech.
- Fixed some indentations and removed some extra comments to maintain consistency.
- Simplified `BlockValidationState` create method by @stringintech [comment](https://github.com/bitcoin/bitcoin/pull/33822#discussion_r2532058549)

stickies-v commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1259,3 +1308,59 @@ int btck_chain_contains(const btck_Chain* chain, const btck_BlockTreeEntry* entr
+    try {
+        stream >> *header;
+    } catch (...) {
+        LogDebug(BCLog::KERNEL, "Block header decode failed.");
```
This should be `LogError`

TheCharlatan commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1017,6 +1036,29 @@ BITCOINKERNEL_API void btck_chainstate_manager_options_destroy(btck_ChainstateMa
+ * @param[in] chainstate_manager Non-null.
+ * @return                       The block tree entry.
+ */
+BITCOINKERNEL_API const btck_BlockTreeEntry* BITCOINKERNEL_WARN_UNUSED_RESULT btck_chainstate_manager_get_best_entry(
```
I think it should be part of this PR. We should not make changes to the API if we don't know how completing simple tasks will look like in the future. If there are doubts about how this should be exposed, this seems like the correct place to voice them and come up with something better. Would you be more on favour if it returned the header directly instead of having to retrieve the header from the entry?

stickies-v replied:
> We should not make changes to the API if we don't know how completing simple tasks will look like in the future.

I agree with that, but I think I'm misunderstanding the purpose of this PR, then. Adding block headers to the API seems like a pretty straightforward addition to the existing API for which imo just the additional introspection ability seems worthwhile, and that's how I was reviewing it. From your messages, it seems like the actual goal is to implement a specific use case, but that doesn't seem to be mentioned anywhere? Can you elaborate on what we're actually trying to achieve here, so people can review and test accordingly? That rationale (and perhaps links to example(s) if they exist) should then also be reflected in the PR description and commit messages.

> Would you be more on favour if it returned the header directly instead of having to retrieve the header from the entry?

I don't think so, that seems wasteful.

yuvicc commented:
Thanks for the review @stickies-v 
- Address usage of `LogError` instead of `LogDebug` [comment](https://github.com/bitcoin/bitcoin/pull/33822#discussion_r2534865290)
- Addressed some nits and removed docstrings.

yuvicc committed:
kernel: Add support for block headers

Introduces btck_BlockHeader type with accessor methods and btck_chainstate_manager_process_block_header() for validating headers without full blocks. Also, adds btck_chainstate_manager_get_best_entry() to query the header with most cumulative proof-of-work.

Co-authored-by: TheCharlatan <seb.kung@gmail.com>

sedited commented on src/kernel/bitcoinkernel_wrapper.h:
```diff
@@ -824,36 +824,50 @@ class KernelNotifications
 };
 
-class BlockValidationState
+template<typename Derived>
```
Nit (clang-tidy): Needs a whitespace after `template`.

sedited reviewed (approved):
ACK 10389bcfcf3eb92a08ff1fafab86db89749520c2

yuvicc commented:
- Addressed some clang-format nits
- Updated the comments to use struct name instead of raw names as suggested by @stickies-v.

sedited reviewed (approved):
re-ACK 30367c1b3e7e94ed59d0a3ee816da14a3ee0424f

alexanderwiederin reviewed (commented):
Concept ACK

sedited commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1253,6 +1271,22 @@ int btck_chainstate_manager_process_block(
+{
+    try {
+        auto& chainman = btck_ChainstateManager::get(chainstate_manager).m_chainman;
+        auto result = chainman->ProcessNewBlockHeaders({&btck_BlockHeader::get(header), 1}, /*min_pow_checked=*/true, btck_BlockValidationState::get(state), /*ppindex=*/nullptr);
```
This is on purpose from my part. I think it is confusing and not helpful to expose. If you look at where it is used, you'll see that all it does is invalidate a header that does not extend the pre-synced best work chain. The only case it is still used in our code is when processing unrequested blocks, but since we're processing headers here, I don't think that is relevant. I also don't think that this should be documented either. The current `min_pow_checked` handling in our validation code is confusing and noisy, and should be refactored and improved in my  opinion.

EDIT: Also note that the same is done for the currently exposed RPC call for processing a header: https://github.com/bitcoin/bitcoin/blob/master/src/rpc/mining.cpp#L1127

stringintech commented on src/kernel/bitcoinkernel_wrapper.h:
```diff
@@ -824,36 +897,50 @@ class KernelNotifications
-    virtual void BlockChecked(Block block, const BlockValidationState state) {}
+    virtual void BlockChecked(Block block, const BlockValidationStateView state) {}
 
     virtual void PowValidBlock(BlockTreeEntry entry, Block block) {}
```
I think we can drop `const`. We haven't enforced it for `BlockTreeEntry` which is also a view. (Also for `BlockChecked` definitions in `bitcoin-chainstate.cpp` and `test_kernel.cpp`)
It seems throughout the wrapper, we haven't stated const-ness also for functions returning views.

stringintech commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1561,6 +1634,92 @@ BITCOINKERNEL_API void btck_block_hash_destroy(btck_BlockHash* block_hash);
+ * @return                          Block Header, or null on error.
+ */
+BITCOINKERNEL_API btck_BlockHeader* BITCOINKERNEL_WARN_UNUSED_RESULT btck_block_header_create(
+    const void* raw_block_header, size_t raw_block_header_len) BITCOINKERNEL_ARG_NONNULL(1);
```
`must be 80` might suggest there is an enforcement of the length somewhere while for example in the `test_kernel.cpp` appending extra bytes to `mainnet_block_1_header` won't break the test and extra bytes will just be discarded. 
Might be helpful to have the enforcement in `btck_block_header_create` or revise the doc.

yuvicc commented on src/kernel/bitcoinkernel.cpp:
```diff
@@ -1238,6 +1271,22 @@ int btck_chainstate_manager_process_block(
+    } catch (std::exception& e) {
+        LogError("Failed to process block header: %s", e.what());
+        return -1;
+    }
```
Correct we don't run into exceptions here but its better to keep for defensive reason. see https://github.com/bitcoin/bitcoin/pull/33822#discussion_r2508041569

stringintech commented on src/bitcoin-chainstate.cpp:
```diff
@@ -57,7 +57,7 @@ class TestValidationInterface : public ValidationInterface
 
     std::optional<std::string> m_expected_valid_block = std::nullopt;
 
-    void BlockChecked(const Block block, const BlockValidationState state) override
```
..

stringintech reviewed (commented):
ACK 4a261e9f

sedited commented on src/kernel/bitcoinkernel.h:
```diff
@@ -1615,6 +1688,92 @@ BITCOINKERNEL_API void btck_block_hash_destroy(btck_BlockHash* block_hash);
+    const btck_BlockHeader* header) BITCOINKERNEL_ARG_NONNULL(1);
+
+/**
+ * @brief Get the previous btck_BlockHash from btck_BlockHeader. The returned hash is unowned and only valid for the lifetime of the btck_BlockHeader.
```
Nit: Can you break this line?

sedited reviewed (approved):
Re-ACK 4a261e9f91289427ceda11cac1acac7bc26102c1

sedited reviewed (approved):
Re-ACK 9356833595f46a6d57d9e5f4a0504f0f8c4256c9

stringintech reviewed (commented):
re-ACK 9356833

janb84 reviewed (commented):
Concept ACK 9356833595f46a6d57d9e5f4a0504f0f8c4256c9

Adds Block header support to the kernel. 

Was able to create a working implementation of the changes in BitcoinKernel.NET.
Found Small Nit during code review

sedited reviewed (approved):
Re-ACK 9a9d797ef6ed8e1b3e876fc93cf1a6395ab270e9

janb84 reviewed (commented):
ACK 9a9d797ef6ed8e1b3e876fc93cf1a6395ab270e9

thanks for incorporating my nit!
